Implement RogueGenocide as options

Zehra opened the pull request:
Compile tested patch which should implement "RogueGenocide" as an option within the "Genocide" plug-in.

Zehra committed:
compile test version of genocide + rogueGeno

Zehra committed:
Update readme to reflect updated plugin

allejo commented:
I'm against using arbitrary numbers for configuration options. I'd prefer having understandable flags like `allowRogue` and `allowSuicide` to be comma-separated as the plug-in's configuration.

```
-loadplugin genocide,allowRogue,allowSuicide
```

This way, we can also allow suicide genocide but not rogue genocide.

Zehra commented:
Would option for "suicide" apply to "teams" as well?
Since this could allow for where selfkills would not effect teams, which where selfkills trigger the effect in both 2.4.x and 2.x.
(However there was a patch provided for 2.99 which would have prevented selfkills from affecting teams, if the "**-noTeamKills**" option were enabled.)

Improved flags will be added next commit, just want to clarify above, as it would introduce changes to game play.

allejo commented:
Blegh I don't want to have so many different configuration options of Genocide. How about these options:

- `rogueAsTeam`
- `disableSuicide` (for everyone)

The default behavior will be to allow suicide ONLY for colored teams. Suicide for rogues will only automatically happen when `rogueAsTeam` is used. When `disableSuicide` is used, it will disable suicide for all teams including rogues.

Thoughts @blast007?

macsforme commented:
A complete implementation would also remove the RogueGenocide plugin from the source as well as (at least) from the makefile system. This change also calls for documentation in the ChangeLog.

Zehra commented:
Will update with both recommendations once consensus on game play is established.
My familiarity with Git is somewhat limited, so possibly a bit of time before next update.
(This way both actions may be done in a single go vs multiple commits for a single action.)

Zehra commented:
Ah, just realized what happened:

Currently with RogueGenocide and Genocide as plug-ins, the following states are possible:
1. Genocide is loaded by itself.
2. Genocide and RogueGenocide is loaded.
3. Genocide and RogueGenocide is loaded with "suicide" option.
No other states are possible:
(Since simply loading RogueGenocide by itself will do nothing.)

Due to this, it would be impossible to have "suicide" enabled, as nothing would occur.(Unless RogueGeno is enabled.)
(In the current version, hence my confusion of if an additional state would be affected, as self-kills triggering geno already effects teams and is not possible to disable it, either with options or "-noTeamKills" setting.)

For a future release, there was a patch which disabled genocide from effecting teams by self-kills, if "-noTeamKills" were enabled.
(However this may be an outdated goal.)

So in theory, there's only "2" options, with one option being a side option of a side option. (With current PR.)
Genocide being default, RogueGenocide being side option of Genocide, with "suicide" being an option of RogueGenocide.

---

Proposed modification results in the following possible states:
1. Genocide default.
2. Genocide with RogueGeno.
3. Genocide with RogueGeno and disable suicide.
4. Genocide with disable suicide.

The only changes which result from it are the following with regards to current and proposed game play is:
_4. Genocide with disable suicide._
This partially implements behavior of an accepted patch for a proposed version of the game and does not in any existing versions.
This change would mainly emulate said patch and be the only differing version of the game play.
(Except in the case of where team kills are enabled and "disable suicide" is enabled" is the main difference from patch.)
Hence proposed modification would only introduce a single game play modification.

Zehra committed:
Removed RogueGenocide as merged into Genocide

Zehra committed:
Update genocide.cpp and removed rogueGenocide from build

Zehra committed:
Oops. (Fixes bug I introduced in last change.)

This would have resulted in incorrect behavior of plugin.

jwmelto commented on plugins/genocide/genocide.cpp:
```diff
@@ -16,29 +16,47 @@ class genocide : public bz_Plugin
 {
     bz_debugMessage(4, "genocide plugin loaded");
-    if (commandLine)
+    if (strlen(cmdLine) != 0)
```
This is a fundamental change in functionality/stability.

One of the dangers of pointers is that they have the potential to be null. The previous code tested for null. Here you assume it's valid, and points to memory with a single null byte. 

Testing for an empty string is pretty pointless.

jwmelto commented on plugins/genocide/genocide.cpp:
```diff
@@ -16,29 +16,47 @@ class genocide : public bz_Plugin
-            rogueGenocide = true;
-        if (conf == 2)
-            allowSuicide = true;
+        if (strcmp(cmdLine, "rogueAsTeam,disableSuicide") == 0)
```
this class of methods are dangerous. Prefer `strncmp` and friends instead. Or, better yet, use `std::string` instead.

jwmelto commented on plugins/genocide/genocide.cpp:
```diff
@@ -16,29 +16,47 @@ class genocide : public bz_Plugin
+        else
+        {
+            bz_debugMessage(0, "Error: Check arguments.");
+        }
```
this block could be simplified with `std::string::find`, as in:

```c++
auto cmd = std::string(cmdLine);
disableSuicide =  (cmd.find("disableSuicide") != cmd.npos);
rogueAsTeam    =  (cmd.find("rogueAsTeam")    != cmd.npos);
```

jwmelto commented on plugins/genocide/genocide.cpp:
```diff
@@ -16,29 +16,47 @@ class genocide : public bz_Plugin
 
-    if (rogueGenocide == true)
-        bz_debugMessage(4, "rogue genocide enabled");
+    if (rogueAsTeam == true)
```
just a personal preference, but testing boolean variables against explicit boolean values is redundant.

Well named variables should be self-explanatory, as in
```c++
if (rogueAsTeam)
   ...
```

jwmelto commented:
by the way, don't just remove from the build; delete the files as well

Zehra committed:
Implement suggestions/recommends and update docs

Zehra committed:
Merge branch 'geno-patch' of https://github.com/Zehra/bzflag into geno-patch

Zehra commented:
Updated with suggestions/recommendations.
I'm not sure, but [I think I deleted the files as well](https://github.com/BZFlag-Dev/bzflag/pull/261/commits/3a61d52c8b1868bbbbd2957a2b10df51d0425b06)? (I might have missed something though.)

Zehra commented on plugins/genocide/genocide.cpp:
```diff
@@ -12,25 +12,37 @@ class genocide : public bz_Plugin
     virtual void Cleanup ( void );
     virtual void Event ( bz_EventData *eventData );
+
+    bool disableSuicide = false;
```
Within the _RogueGenocide_ plugin, it was done as such.
Any recommendations? Since I've seen both **private** and **protected** used in a few other plugins
For instance, _playHistoryTracker_ uses **protected**, _regFlag_ uses **private** and _airspawn_ uses **public.**

Zehra committed:
ChangeLog addition and minor update to geno plugin

allejo commented on plugins/genocide/genocide.cpp:
```diff
@@ -53,64 +66,67 @@ void genocide::Event(bz_EventData *eventData)
+        if (dieData->team == eRogueTeam)
         {
-            // if the tank killed was a rogue, kill all rogues.
-            bz_APIIntList *playerList = bz_newIntList();
+            for (unsigned int i = 0; i < playerList->size(); i++)
+            {
+                int targetID = (*playerList)[i];
+                bz_BasePlayerRecord *playRec = bz_getPlayerByIndex (targetID);
+                if (!playRec)
+                    continue;
 
-            bz_getPlayerIndexList(playerList);
+                // the sucker is a spawned rogue, kill him.  This generates another death event,
+                // so if you kill another rogue with geno while you are a rogue you end up dead too.
+                // and you get both messages (victim and be careful)
+                if (playRec->spawned && playRec->team == eRogueTeam)
+                {
+                    bz_killPlayer(targetID, false, dieData->killerID, "G");
+                    bz_sendTextMessage(BZ_SERVER, targetID, "You were a victim of Rogue Genocide");
 
+                    // oops, I ended up killing myself (directly or indirectly) with Genocide!
+                    if (targetID == dieData->killerID)
+                        bz_sendTextMessage(BZ_SERVER, targetID, "You should be more careful with Genocide!");
+                }
+
+                bz_freePlayerRecord(playRec);
+            }
+        }
```
- These are identical loops, there's no need to split them for rogues only; combine the two.
- Is there a reason `eGenocideEffect` doesn't work for rogue tanks? Skimming through the source code, there doesn't seem to be anything special server-side or client-side disabling `eGenocideEffect` from affecting rogues. Can you test this behavior to make sure `eGenocideEffect` works or doesn't work?
- The "be careful message" shouldn't just apply to rogue genocide, that should apply to genocide in general, but should only appear if you're **not** rogue.
